Fix image cropping for non-admins

swissspidy opened the pull request:
## Context

<!-- What do we want to achieve with this PR? Why did we write this code? -->

Users with editor role were not able to crop poster images

## Summary

<!-- A brief description of what this PR does. -->

Fixes an issue with image cropping for users without the `customize` capability.

Directly using `CustomizeImageCropper` will set the `wp_customize` param which only works for admins, but in our case the cropping should also work for non-admins.

## Relevant Technical Choices

<!-- Please describe your changes. -->

Adds a custom cropper similar to `CustomizeImageCropper`.

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

## User-facing changes

<!--
Please describe your changes.
Include before/after screenshots or a short video.
-->

N/A

## Testing Instructions

<!--
How can the changes in this PR be verified?
Please provide step-by-step instructions how to reproduce the issue, if applicable.
Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->

<!-- ignore-task-list-start -->
- [ ] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

This PR can be tested by following these steps:

1. Log in as an editor
2. Create a new story
3. Try to set a poster image or publisher logo which requires cropping
4. Verify that cropping works.


## Reviews

### Does this PR have a security-related impact?

<!-- Examples: new APIs, changes to KSES, etc.  -->

### Does this PR change what data or activity we track or use?

<!-- Examples: changes to telemetry, new third-party APIs -->

### Does this PR have a legal-related impact?

<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.

NOTE: One reference per line!

Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #8912

swissspidy committed:
Fix image cropping for non-admins

swissspidy committed:
Update fixtures

swissspidy committed:
Lint fix

swissspidy committed:
Update phpdoc

swissspidy committed:
Lint fixes

spacedmonkey commented on packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js:
```diff
@@ -0,0 +1,66 @@
+  },
+});
+
+window.wp.media.controller.WordPressImageCropper = WordPressImageCropper;
```
Why set this to the window? Why not just export as default like everywhere else?

swissspidy replied:
I was debugging the Karma tests. Will probably revert again.

spacedmonkey replied:
> I was debugging the Karma tests. Will probably revert again.

Can you explain this a little more?

swissspidy replied:
Well the Karma tests are failing due to some globals being undefined and I was trying to debug why this happens and to make the Karma tests pass.

If you have a hunch what might be the issue, feel free to jump in and help fix the tests.

spacedmonkey commented on packages/story-editor/src/components/mediaPicker/useMediaPicker.js:
```diff
@@ -196,7 +197,7 @@ export default function useMediaPicker({
 
       // Create the media frame.
-      const fileFrame = global.wp.media({
+      const fileFrame = window.wp.media({
```
What does this change effect?

swissspidy replied:
Nothing. They're exactly the same. But webpack replaces `global` with `window` when bundling.

And in webpack v5 it actually deprecates using `global`. You can learn more about that in #5792 

I was just fixing this while being at it.

spacedmonkey commented on packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js:
```diff
@@ -0,0 +1,66 @@
+   * Uses crop measurements when flexible in both directions.
+   * Constrains flexible side based on image ratio and size of the fixed side.
+   *
+   * @since 4.3.0
```
```suggestion
```

spacedmonkey reviewed (changes requested):
Some nitpics, but otherwise looking good.

spacedmonkey commented on packages/story-editor/src/components/checklist/karma/checklist.karma.js:
```diff
@@ -402,10 +402,10 @@ describe('Checklist integration - Card visibility', () => {
 
     class Library {}
 
-    class CustomizeImageCropper {}
```
```
class Cropper {
    extend(){}
}

media.controller = {
      Cropper,
      Library,
};
```

I think this is issue. Remember we call `new window.wp.media.controller.WordPressImageCropper({` So cropper needs to be a class.

spacedmonkey replied:
I pushed it, let's see what happens.

spacedmonkey replied:
https://github.com/google/web-stories-wp/pull/8924/commits/ca022fd57e3a45a5ef9e4a718353a917bdc8f732

swissspidy replied:
Can you reply to the CLA bot too?

Looks like there are still `Uncaught TypeError: Cannot read properties of undefined (reading 'media')` errors unfortunately.

spacedmonkey committed:
Use class.

spacedmonkey committed:
Export as variable.

spacedmonkey committed:
Remove since.

spacedmonkey committed:
Keep inline with other mock.

spacedmonkey committed:
Merge branch 'main' into fix/editor-image-crop

swissspidy committed:
Return class in `extend()`

swissspidy committed:
Use optional chaining

Needed  since `wp.media` is not available on the Dashboard page but somehow this class is being included in the bundle.

swissspidy committed:
Merge branch 'main' into fix/editor-image-crop

swissspidy committed:
Add safeguard
